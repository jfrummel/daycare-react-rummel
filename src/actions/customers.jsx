import { v4 as uuidv4 } from 'uuid';

//  Customers action generators

export const addCustomer = ({ firstName = "", lastName = "", numOfChildren = 0, hourlyRate = 0 } = {}) => ({
    type: "ADD_CUSTOMER",
    customer: {
        id: uuidv4(),
        firstName,
        lastName,
        numOfChildren,
        hourlyRate
    }
});

export const removeCustomer = ({ id } = {}) => ({
    type: "REMOVE_CUSTOMER",
    id
});

export const editCustomer = (id, updates) => ({
    type: "EDIT_CUSTOMER",
    id, updates
});