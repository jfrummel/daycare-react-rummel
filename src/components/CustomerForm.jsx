import React, { Component } from 'react';



class CustomerForm extends Component {
    constructor(props) {
        super(props);

        this.state = {
            firstName: props.customer ? props.customer.firstName : "",
            lastName: props.customer ? props.customer.lastName : "",
            numOfChildren: props.customer ? props.customer.numOfChildren : 0,
            hourlyRate: props.customer ? (props.customer.hourlyRate.toString()) : "",
            error: ""
        }
    }

    onFirstNameChange = (e) => {
        const firstName = e.target.value;
        this.setState(() => ({ firstName }));
    };

    onLastNameChange = (e) => {
        const lastName = e.target.value;
        this.setState(() => ({ lastName }));
    };

    onNumChildrenChange = (e) => {
        const numOfChildren = e.target.value;
        this.setState(() => ({ numOfChildren }));
    };

    onHourlyRateChange = (e) => {
        const hourlyRate = e.target.value;
        if (!hourlyRate || hourlyRate.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ hourlyRate }));
        }
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.firstName || !this.state.lastName) {
            this.setState(() => ({ error: "First and Last Name Required" }))
        } else {
            this.setState(() => ({ error: "" }));
            this.props.onSubmit({
                firstName: this.state.firstName,
                lastName: this.state.lastName,
                numOfChildren: this.state.numOfChildren,
                hourlyRate: parseFloat(this.state.hourlyRate)
            });
        };
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input
                    type="text"
                    value={this.state.firstName}
                    placeholder="First Name"
                    onChange={this.onFirstNameChange}
                    autoFocus
                />
                <input
                    type="text"
                    value={this.state.lastName}
                    placeholder="Last Name"
                    onChange={this.onLastNameChange}
                />
                <input
                    type="number"
                    value={this.state.numOfChildren}
                    placeholder="Number of Children"
                    onChange={this.onNumChildrenChange}
                />
                <input
                    type="text"
                    value={this.state.hourlyRate}
                    placeholder="Hourly Rate"
                    onChange={this.onHourlyRateChange}
                />
                <button>{this.props.buttonText}</button>
            </form>
        );
    }
}

export default CustomerForm;