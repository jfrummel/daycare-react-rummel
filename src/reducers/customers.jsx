//  customers default state

const customerDefaultState = [];

const customerReducer = (state = customerDefaultState, action) => {
    switch (action.type) {
        case "ADD_CUSTOMER":
            return [...state, action.customer];
        case "REMOVE_CUSTOMER":
            return state.filter(({ id }) => id !== action.id);
        case "EDIT_CUSTOMER":
            return state.map((customer) => {
                if (customer.id === action.id) {
                    return { ...customer, ...action.updates };
                } else {
                    return customer
                }
            });
        default:
            return state;
    }
};

export default customerReducer;