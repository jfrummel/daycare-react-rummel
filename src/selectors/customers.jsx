
export const getVisibleCustomers = (customers, { text }) => {
    return customers.filter((customer) => {
        const textMatch = customer.lastName.toLowerCase().includes(text.toLowerCase()) ||
            customer.firstName.toLowerCase().includes(text.toLowerCase());

        return textMatch;
    });
};