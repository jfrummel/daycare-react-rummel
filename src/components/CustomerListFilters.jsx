import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';

class CustomerListFilters extends Component {

    onTextChange = (e) => {
        const text = e.target.value;
        this.props.dispatch(setTextFilter(text));
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

export default connect(mapStateToProps)(CustomerListFilters);