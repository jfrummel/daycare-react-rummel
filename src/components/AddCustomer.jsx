import React, { Component } from 'react';
import CustomerForm from './CustomerForm';
import { connect } from 'react-redux';
import { addCustomer } from '../actions/customers';

export class AddCustomer extends Component {

    onSubmit = (customer) => {
        this.props.addCustomer(customer);
        this.props.history.push('/customer_list');
    };

    render() {
        return (
            <div>
                <h2>Add a new customer</h2>
                <CustomerForm
                    buttonText="Add Customer"
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}




const mapDispatchToProps = (dispatch) => ({
    addCustomer: (customer) => dispatch(addCustomer(customer))
});

export default connect(undefined, mapDispatchToProps)(AddCustomer);