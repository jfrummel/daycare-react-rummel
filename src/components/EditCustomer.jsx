import React, { Component } from 'react';
import { connect } from 'react-redux';
import { editCustomer, removeCustomer } from '../actions/customers';
import CustomerForm from './CustomerForm';

export class EditCustomer extends Component {

    onSubmit = (customer) => {
        this.props.editCustomer(this.props.customer.id, customer);
        this.props.history.push("/customer_list");
    };

    onRemove = () => {
        this.props.removeCustomer({ id: this.props.customer.id });
        this.props.history.push("/customer_list");
    }

    render() {
        return (
            <div>
                <h3>Edit Customer</h3>
                <CustomerForm
                    customer={this.props.customer}
                    buttonText="Save Expense"
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onRemove}
                >
                    Remove
        </button>
            </div>
        );
    }
};

const mapDispatchToProps = (dispatch, props) => ({
    editCustomer: (id, customer) => dispatch(editCustomer(id, customer)),
    removeCustomer: (customer) => dispatch(removeCustomer(customer))
});

const mapStateToProps = (state, props) => {
    return {
        customer: state.customers.find((customer) => customer.id === props.match.params.id)
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditCustomer);