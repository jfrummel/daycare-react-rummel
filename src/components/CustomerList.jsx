import React from 'react';
import { connect } from 'react-redux';
import { getVisibleCustomers } from '../selectors/customers';
import CustomerListFilters from './CustomerListFilters';
import CustomerListItem from './CustomerListItem';

export const CustomerList = (props) => (
    <div>
        <CustomerListFilters />
        {props.customers.length === 0 ? <h3>No Customers found</h3> :
            <ul>
                {props.customers.map((customer) => {
                    return <CustomerListItem key={customer.id} customer={customer} />
                })}
            </ul>
        }

    </div>
);

const mapStateToProps = (state) => ({
    customers: getVisibleCustomers(state.customers, state.filters)
});

export default connect(mapStateToProps)(CustomerList);