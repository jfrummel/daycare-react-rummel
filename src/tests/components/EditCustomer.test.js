import React from 'react';
import { shallow } from 'enzyme';
import { EditCustomer } from '../../components/EditCustomer';
import { customers } from '../fixtures/customers';

let editCustomer, removeCustomer, wrapper, history;

beforeEach(() => {
    editCustomer = jest.fn();
    removeCustomer = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditCustomer
            history={history}
            editCustomer={editCustomer}
            removeCustomer={removeCustomer}
            customer={customers[1]}
        />)
});

test("renders edit customer component correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("handles edit customer correctly", () => {
    wrapper.find('CustomerForm').prop('onSubmit')(customers[1]);
    expect(history.push).toHaveBeenLastCalledWith('/customer_list');
    expect(editCustomer).toHaveBeenLastCalledWith(customers[1].id, customers[1]);
});

test("handles remove customer correctly", () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/customer_list');
    expect(removeCustomer).toHaveBeenLastCalledWith({ id: customers[1].id });
});

