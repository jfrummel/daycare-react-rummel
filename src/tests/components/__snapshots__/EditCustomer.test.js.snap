// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders edit customer component correctly 1`] = `
<div>
  <h3>
    Edit Customer
  </h3>
  <CustomerForm
    buttonText="Save Expense"
    customer={
      Object {
        "firstName": "Lauren",
        "hourlyRate": 3,
        "id": "2",
        "lastName": "Victoria",
        "numOfChildren": 7,
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
